tests/unittests/HyphenTest.cpp

```diff
@@ -0,0 +1,322 @@
+    "treat-ment",
+    "va-ri-ous",
+    "value",
+    "vi-o-lence",
+};
+
+TEST(HyphenTest, network) {
+  // Convert the training data to word windows and labels.
+  vector<string> words;
+  vector<bool> hyphens;
+  for (auto *hword : TrainingData) {
+    dehyphenate(hword, words, hyphens);
+  }
+
+  // This depends on the training data, of course.
+  const size_t numSamples = 566;
+  ASSERT_EQ(hyphens.size(), numSamples);
+  ASSERT_EQ(words.size(), numSamples);
+
+  // Randomly shuffle the training data.
+  // This is required for stochastic gradient descent training.
+  for (size_t i = 0; i != numSamples; i++) {
+    size_t j = nextRandInt(0, numSamples - 1);
+    std::swap(words[i], words[j]);
+    std::swap(hyphens[i], hyphens[j]);
+  }
+
+  // Convert words and hyphens to a tensor representation.
+  Tensor inputs(ElemKind::FloatTy, {numSamples, 6, 27});
+  Tensor outputs(ElemKind::FloatTy, {numSamples, 1});
+  auto inputHandle = inputs.getHandle<float>();
+  auto outputHandle = outputs.getHandle<float>();
+  for (size_t i = 0; i != numSamples; i++) {
+    mapLetterWindow(words[i], i, inputHandle);
+    outputHandle.at({i, 0}) = hyphens[i];
+  }
+
+  // Now build the network:
+  //
+  // Input: 6 x 27.
+  // Hidden: 16 fully connected, RELU output.
+  // Output: 1 fully connected, sigmoid output for logistic regression.
+  const size_t batchSize = numSamples;
+  ExecutionEngine EE(BackendKind::CPU);
+  auto &mod = EE.getModule();
+  Function *Infer = mod.createFunction("infer");
+
+  auto *InputVar =
+      mod.createVariable(ElemKind::FloatTy, {batchSize, 6, 27}, "input",
+                         VisibilityKind::Public, Variable::TrainKind::None);
+  auto *OutputVar =
+      mod.createVariable(ElemKind::FloatTy, {batchSize, 1}, "output",
+                         VisibilityKind::Public, Variable::TrainKind::None);
+
+  auto *HiddenFC = Infer->createFullyConnected("hidden_fc", InputVar, 16);
+  auto *Hidden = Infer->createRELU("hidden", HiddenFC);
+  auto *OutputFC = Infer->createFullyConnected("output_fc", Hidden, 1);
+  auto *Sig = Infer->createSigmoid("sig", OutputFC);
+  auto *Output = Infer->createRegression("regr", Sig, OutputVar);
+  auto *Result = Infer->createSave("result", Output);
+
+  // Generate a gradient network and use it for training.
+  Function *Train = glow::differentiate(Infer, EE.getConfig());
+  EE.compile(CompilationMode::Train, Train);
+  EE.runBatch(200, {InputVar, OutputVar}, {&inputs, &outputs});
+
+  // Now test inference on the trained network.
+  // Note that we have probably overfitted the data.
+  EE.compile(CompilationMode::Infer, Infer);
+  EE.run({InputVar}, {&inputs});
+  auto outHandle = Result->getVariable()->getHandle<>();
+  size_t good = 0, bad = 0;
+  for (size_t i = 0; i != numSamples; i++) {
+    float value = outHandle.at({i, 0});
+    if ((value > 0.5) == hyphens[i]) {
+      good++;
+    } else {
+      bad++;
+    }
+  }
+  EXPECT_GT(good, numSamples / 3);
+  EXPECT_LT(bad, 5);
```

Comment:
What are values of `good` and `bad` that you see? Can we expect 100% accuracy?

Reply:
I get `bad=2` out of 566 samples. It is possible that they are inconsistent samples where we require different outputs for the same input. I'll check. Without inconsistencies, this network should be able to overfit such a small dataset.